Nodejs/Product/TestAdapter/TestExecutorWorker.cs

```diff
@@ -0,0 +1,390 @@
+            VisualStudioApp.LogTelemetry(testCount, nodeVersion, isDebugging, testFramework);
+        }
+
+        private void AttachDebugger(int vsProcessId, int port, Version nodeVersion)
+        {
+            try
+            {
+                if (nodeVersion >= Node8Version)
+                {
+                    VisualStudioApp.AttachToProcessNode2DebugAdapter(vsProcessId, port);
+                }
+                else
+                {
+                    //the '#ping=0' is a special flag to tell VS node debugger not to connect to the port,
+                    //because a connection carries the consequence of setting off --debug-brk, and breakpoints will be missed.
+                    var qualifierUri = string.Format("tcp://localhost:{0}#ping=0", port);
+                    while (!VisualStudioApp.AttachToProcess(this.nodeProcess.Process, vsProcessId, NodejsRemoteDebugPortSupplierUnsecuredId, qualifierUri))
+                    {
+                        if (this.nodeProcess.Wait(TimeSpan.FromMilliseconds(500)))
+                        {
+                            break;
+                        }
+                    }
+                }
+#if DEBUG
+            }
+            catch (COMException ex)
+            {
+                this.frameworkHandle.SendMessage(TestMessageLevel.Error, "Error occurred connecting to debuggee.");
+                this.frameworkHandle.SendMessage(TestMessageLevel.Error, ex.ToString());
+                KillNodeProcess();
+            }
+#else
+            }
+            catch (COMException)
+            {
+                frameworkHandle.SendMessage(TestMessageLevel.Error, "Error occurred connecting to debuggee.");
+                KillNodeProcess();
+            }
+#endif
+        }
+
+
+        private static int GetFreePort()
+        {
+            return Enumerable.Range(new Random().Next(49152, 65536), 60000).Except(
+                from connection in IPGlobalProperties.GetIPGlobalProperties().GetActiveTcpConnections()
+                select connection.LocalEndPoint.Port
+            ).First();
+        }
+
+        private static TestFramework.ArgumentsToRunTests GetInterpreterArgs(TestCase test, string workingDir, string projectRootDir)
+        {
+            var testFile = test.GetPropertyValue(JavaScriptTestCaseProperties.TestFile, defaultValue: test.CodeFilePath);
+            var testFramework = test.GetPropertyValue<string>(JavaScriptTestCaseProperties.TestFramework, defaultValue: null);
+            return FrameworkDiscoverer.Instance.Get(testFramework).GetArgumentsToRunTests(test.DisplayName, testFile, workingDir, projectRootDir);
+        }
+
+        private static string GetDebugArgs(Version nodeVersion, out int port)
+        {
+            port = GetFreePort();
+
+            return nodeVersion >= Node8Version ? $"--inspect-brk={port}" : $"--debug-brk={port}";
+        }
+
+        private void KillNodeProcess()
+        {
+            this.nodeProcess?.Kill();
```

Comment:
>this.nodeProcess?.Kill(); [](start = 12, length = 25)

Does this wait for it to exit?  Should it?

Reply:
It doesn't, and it shouldn't since we only call this for cancel.. I'll move it to the cancel method